api/lib/opentelemetry/context.rb

```diff
@@ -29,23 +30,45 @@ def current
         Thread.current[KEY] ||= ROOT
       end
 
-      # Sets the current context
-      #
-      # @param [Context] ctx The context to be made active
+      # @api private
       def current=(ctx)
         Thread.current[KEY] = ctx
       end
 
+      # Returns a token to be used with the matching call to detach
+      #
+      # @param [Context] context The new context
+      # @return [Integer] A token to be used when detaching
+      def attach(context)
+        prev = current
+        self.current = context
+        stack.push(prev)
+        stack.size
+      end
+
+      # Restores the previous context, if a token is supplied it will
+      # be used to check if the call to detach is balanced with
+      # the original attach call. A warning is logged if the
+      # calls are unbalanced.
+      #
+      # @param [Integer] token The token provided by the matching call to attach
+      def detach(token = nil)
+        OpenTelemetry.logger.warn 'Calls to detach should match corresponding calls to attach' if token && token != stack.size
```

Comment:
```suggestion
      def detach(token)
        OpenTelemetry.logger.warn 'Calls to detach should match corresponding calls to attach' if token != stack.size
```

Reply:
Also, do you think there's value in logging the stack at this point, or at least the caller?